

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genemap.mappers.mgi &mdash; Genemap 0.2.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="Genemap 0.2.0 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> Genemap
          

          
          </a>

          
            
            
              <div class="version">
                0.2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Genemap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line.html">Command line</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../history.html">History</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Genemap</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>genemap.mappers.mgi</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for genemap.mappers.mgi</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1"># pylint: disable=wildcard-import,redefined-builtin,unused-wildcard-import</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">division</span><span class="p">,</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="k">import</span> <span class="o">*</span>

<span class="c1"># pylint: enable=wildcard-import,redefined-builtin,unused-wildcard-import</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">StringIO</span> <span class="k">import</span> <span class="n">StringIO</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">io</span> <span class="k">import</span> <span class="n">StringIO</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">import</span> <span class="nn">requests_cache</span>

<span class="kn">from</span> <span class="nn">.base</span> <span class="k">import</span> <span class="n">Mapper</span><span class="p">,</span> <span class="n">CommandLineMixin</span><span class="p">,</span> <span class="n">register_mapper</span>

<span class="n">MAP_IDS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span> <span class="s1">&#39;entrez&#39;</span><span class="p">}</span>
<span class="n">MAP_URL</span> <span class="o">=</span> <span class="s1">&#39;http://www.informatics.jax.org/downloads/reports/HOM_AllOrganism.rpt&#39;</span>

<span class="n">MAP_DTYPES</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;HomoloGene ID&#39;</span><span class="p">:</span> <span class="s1">&#39;str&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Common Organism Name&#39;</span><span class="p">:</span> <span class="s1">&#39;str&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Symbol&#39;</span><span class="p">:</span> <span class="s1">&#39;str&#39;</span><span class="p">,</span>
    <span class="s1">&#39;EntrezGene ID&#39;</span><span class="p">:</span> <span class="s1">&#39;str&#39;</span>
<span class="p">}</span>

<span class="n">requests_cache</span><span class="o">.</span><span class="n">install_cache</span><span class="p">(</span><span class="s1">&#39;.genemap&#39;</span><span class="p">)</span>


<div class="viewcode-block" id="MgiMapper"><a class="viewcode-back" href="../../../api.html#genemap.mappers.MgiMapper">[docs]</a><span class="k">class</span> <span class="nc">MgiMapper</span><span class="p">(</span><span class="n">CommandLineMixin</span><span class="p">,</span> <span class="n">Mapper</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;MGI mapper class.</span>

<span class="sd">    Maps IDs using the homology table provided by MGI. Supports mapping between</span>
<span class="sd">    gene symbols and Entrez identifers and between several specifies</span>
<span class="sd">    (using orthologous genes).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    from_type : str</span>
<span class="sd">        The source identifier type. Can be either &#39;symbol&#39; or &#39;entrez&#39;.</span>
<span class="sd">    to_type : str</span>
<span class="sd">        The target identifier type (see ``from_type`` for more details).</span>
<span class="sd">    drop_duplicates : str</span>
<span class="sd">        How to handle duplicates. If &#39;both&#39;, then entries with duplicates in</span>
<span class="sd">        either the source/target columns are dropped from the mapping. If</span>
<span class="sd">        &#39;mto&#39; (many-to-one), then only duplicates in the source column are</span>
<span class="sd">        dropped. If &#39;otm&#39;, then only duplicates in the target column are</span>
<span class="sd">        dropped. Finally, if &#39;none&#39;, no duplicates are removed from the mapping.</span>
<span class="sd">    from_organism : str</span>
<span class="sd">        Name of the source organism.</span>
<span class="sd">    to_organism : str</span>
<span class="sd">        Name of the target organism. If not given, no mapping between</span>
<span class="sd">        organisms is performed.</span>
<span class="sd">    map_url : str</span>
<span class="sd">        The URL to use to fetch the mapping table from MGI.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">from_type</span><span class="p">,</span>
                 <span class="n">to_type</span><span class="p">,</span>
                 <span class="n">drop_duplicates</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>
                 <span class="n">from_organism</span><span class="o">=</span><span class="s1">&#39;mouse&#39;</span><span class="p">,</span>
                 <span class="n">to_organism</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">map_url</span><span class="o">=</span><span class="n">MAP_URL</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">drop_duplicates</span><span class="o">=</span><span class="n">drop_duplicates</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">from_type</span> <span class="o">==</span> <span class="n">to_type</span> <span class="ow">and</span> <span class="p">(</span><span class="n">from_organism</span> <span class="o">==</span> <span class="n">to_organism</span> <span class="ow">or</span>
                                     <span class="n">to_organism</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Cannot map between identical identifier types &#39;</span>
                             <span class="s1">&#39;within the same organism&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_from_type</span> <span class="o">=</span> <span class="n">from_type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_to_type</span> <span class="o">=</span> <span class="n">to_type</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_from_organism</span> <span class="o">=</span> <span class="n">from_organism</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_to_organism</span> <span class="o">=</span> <span class="n">to_organism</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_map_url</span> <span class="o">=</span> <span class="n">map_url</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">configure_parser</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">parser</span><span class="p">):</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--from_type&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="n">MAP_IDS</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--to_type&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="n">MAP_IDS</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--from_organism&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;mouse&#39;</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--to_organism&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--drop_duplicates&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--map_url&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">MAP_URL</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_args</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">from_type</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">from_type</span><span class="p">,</span>
                   <span class="n">to_type</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">to_type</span><span class="p">,</span>
                   <span class="n">from_organism</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">from_organism</span><span class="p">,</span>
                   <span class="n">to_organism</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">to_organism</span><span class="p">,</span>
                   <span class="n">drop_duplicates</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">,</span>
                   <span class="n">map_url</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">map_url</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_fetch_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Fetch and read MGI data.</span>
        <span class="n">req</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_map_url</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">req</span><span class="o">.</span><span class="n">text</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">MAP_DTYPES</span><span class="p">)</span>

        <span class="c1"># Subset and tidy column names.</span>
        <span class="n">column_map</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;HomoloGene ID&#39;</span><span class="p">:</span> <span class="s1">&#39;id&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Common Organism Name&#39;</span><span class="p">:</span> <span class="s1">&#39;organism&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Symbol&#39;</span><span class="p">:</span> <span class="s1">&#39;symbol&#39;</span><span class="p">,</span>
            <span class="s1">&#39;EntrezGene ID&#39;</span><span class="p">:</span> <span class="s1">&#39;entrez&#39;</span>
        <span class="p">}</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">column_map</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">column_map</span><span class="p">)</span>

        <span class="c1"># Extract main organism name (before comma).</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span>
            <span class="sa">r</span><span class="s1">&#39;(\w+),?&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;entrez&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;entrez&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

        <span class="c1"># Check if organisms are known.</span>
        <span class="n">organisms</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">])</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_organism</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">organisms</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unknown from organism </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_from_organism</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_organism</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_organism</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">organisms</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unknown to organism </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_to_organism</span><span class="p">))</span>

            <span class="c1"># Extract rows belong to each organism.</span>
            <span class="n">from_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_organism</span><span class="p">]</span>
            <span class="n">to_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_organism</span><span class="p">]</span>

            <span class="c1"># Merge into single frame based on the &#39;HomoloGene ID&#39;.</span>
            <span class="n">suffixes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_organism</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_organism</span><span class="p">]</span>
            <span class="n">mapping</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                <span class="n">from_data</span><span class="p">,</span> <span class="n">to_data</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="n">suffixes</span><span class="p">)</span>

            <span class="c1"># Extract relevant columns.</span>
            <span class="n">mapping</span> <span class="o">=</span> <span class="n">mapping</span><span class="p">[[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_from_type</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_organism</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_to_type</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_organism</span>
            <span class="p">]]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Extract rows/columns belonging to organisms + types.</span>
            <span class="n">mapping</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_organism</span><span class="p">,</span>
                               <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_from_type</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_type</span><span class="p">]]</span>

        <span class="k">return</span> <span class="n">mapping</span></div>


<span class="n">register_mapper</span><span class="p">(</span><span class="s1">&#39;mgi&#39;</span><span class="p">,</span> <span class="n">MgiMapper</span><span class="p">)</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.2.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>